src/accsr/remote_storage.py

```diff
@@ -214,61 +218,134 @@ def pull(
             (which is discouraged).
         :return: list of objects referring to all downloaded files
         """
+
         local_base_dir = os.path.abspath(local_base_dir)
         if convert_to_linux_path:
             remote_path = remote_path.replace("\\", "/")
-        full_remote_path = self._full_remote_path(remote_path)
-        remote_objects: List[RemoteObjectProtocol] = list(
-            self.bucket.list_objects(full_remote_path)
-        )
-        if len(remote_objects) == 0:
-            log.warning(
-                f"No such remote file or directory: {full_remote_path}. Not pulling anything"
-            )
-            return []
 
-        def maybe_get_destination_path(obj: RemoteObjectProtocol):
-            # Due to a possible bug in libcloud or storage providers, directories may be listed in remote objects.
-            # We filter them out by checking for size
-            if obj.size == 0:
-                log.info(f"Skipping download of {obj.name} with size zero.")
-                return
+        summary = self.collect_pull_summary(remote_path, local_base_dir, path_regex)
 
-            if self._listed_due_to_name_collision(full_remote_path, obj):
-                log.debug(
-                    f"Skipping download of {obj.name}. "
-                    f"It was listed due to name collision and should not be pulled"
-                )
-                return
+        valid_remote_objects = summary["new_files"]
+        existing_eq_md5 = summary["existing_eq_md5"]
+        existing_neq_md5 = summary["existing_neq_md5"]
+        existing = existing_neq_md5 + existing_eq_md5
 
-            relative_obj_path = self._get_relative_remote_path(obj)
-            if path_regex is not None:
-                if not path_regex.match(relative_obj_path):
-                    log.info(f"Skipping {relative_obj_path} due to regex {path_regex}")
-                return
-            return os.path.join(local_base_dir, relative_obj_path)
+        download_list = valid_remote_objects
+        if overwrite_existing:
+            download_list = download_list + existing_neq_md5
+        download_size = sum([obj.size for obj in download_list])
 
-        downloaded_objects = []
-        for remote_obj in remote_objects:
-            destination_path = maybe_get_destination_path(remote_obj)
-            if destination_path is None:
-                continue
-
-            was_downloaded = self._pull_object(
-                remote_obj,
-                destination_path,
-                overwrite_existing=overwrite_existing,
+        if not force and len(existing) > 0:
+            log.error(
+                "Found existing files. Not pulling anything. To skip all existing files use force=True."
```

Comment:
as commented above, we don't need force and overwrite_existing. Existing files with equal hash sums will almost always be found and should never lead to errors. The error is raised only if force is False and there exist files that would be overwritten by the pull.

Note that it is probably unnecessary to log something on error if you raise an exception anyway - the whole message should be contained in the exception then